service/src/main/java/org/cbioportal/genome_nexus/service/internal/VerifiedGenomicLocationAnnotationServiceImpl.java

```diff
@@ -0,0 +1,200 @@
+        throws VariantAnnotationNotFoundException, VariantAnnotationWebServiceException
+    {
+        VariantAnnotation annotation = genomicLocationAnnotationService.getAnnotation(genomicLocation);
+        VariantAnnotation verifiedAnnotation = verifyOrFailAnnotation(annotation);
+        return verifiedAnnotation;
+    }
+
+    @Override
+    public VariantAnnotation getAnnotation(String genomicLocation)
+        throws VariantAnnotationNotFoundException, VariantAnnotationWebServiceException
+    {
+        VariantAnnotation annotation = genomicLocationAnnotationService.getAnnotation(genomicLocation);
+        VariantAnnotation verifiedAnnotation = verifyOrFailAnnotation(annotation);
+        return verifiedAnnotation;
+    }
+
+    @Override
+    public List<VariantAnnotation> getAnnotations(List<GenomicLocation> genomicLocations)
+    {
+        List<VariantAnnotation> annotations = genomicLocationAnnotationService.getAnnotations(genomicLocations);
+        for (int index = 0; index < annotations.size(); index = index + 1) {
+            VariantAnnotation annotation = annotations.get(index);
+            VariantAnnotation verifiedAnnotation = verifyOrFailAnnotation(annotation);
+            annotations.set(index, verifiedAnnotation);
+        }
+        return annotations;
+    }
+
+    @Override
+    public VariantAnnotation getAnnotation(String genomicLocation,
+                                           String isoformOverrideSource,
+                                           Map<String, String> token,
+                                           List<String> fields)
+        throws VariantAnnotationWebServiceException, VariantAnnotationNotFoundException
+    {
+        VariantAnnotation annotation = genomicLocationAnnotationService.getAnnotation(genomicLocation, isoformOverrideSource, token, fields);
+        VariantAnnotation verifiedAnnotation = verifyOrFailAnnotation(annotation);
+        return verifiedAnnotation;
+    }
+
+    @Override
+    public List<VariantAnnotation> getAnnotations(List<GenomicLocation> genomicLocations,
+                                                                    String isoformOverrideSource,
+                                                                    Map<String, String> token,
+                                                                    List<String> fields)
+    {
+        List<VariantAnnotation> annotations = genomicLocationAnnotationService.getAnnotations(genomicLocations, isoformOverrideSource, token, fields);
+        for (int index = 0; index < annotations.size(); index = index + 1) {
+            VariantAnnotation annotation = annotations.get(index);
+            VariantAnnotation verifiedAnnotation = verifyOrFailAnnotation(annotation);
+            annotations.set(index, verifiedAnnotation);
+        }
+        return annotations;
+    }
+
+    private VariantAnnotation verifyOrFailAnnotation(VariantAnnotation annotation)
```

Comment:
quick question here: would this handle the case where the original input query's reference allele has a common prefix with the tumor allele?

I guess what I'm wondering is whether an adjusted reference allele from the input query would still match the reference allele resolved from the genome? if the coordinates were adjusted also then I think this should probably be taken care of already but just wanted to double check

Reply:
actually, is this even applicable in this context since "Verified" annotation service implementations appear to require a local running vep service? or am i mistaken?

Reply:
First clarification : the Verified annotation services always run, regardless of whether genome-nexus is using a local deployment of vep (genome-nexus-vep) or is using the public ensembl VEP API for remote prediction generation. When using public ensembl predictions, all queries are converted into HGVS format and are submitted to the /vep/human/hgvs/{VARIANT_GOES_HERE} endpoint at ensembl (or wherever the deployer has set the property `vep.url` to point to. This seems to always use the GET method endpoint for fetching from Ensembl (VEPDataFetcher). When using a local VEP deployment for predictions, all queries are converted into REST-like Region format for processing by the VEP command line tool in the genome-nexus-vep image. Note that in both cases, all reference allele information is lost : Region format does not include reference allele strings, and the NotationConverter logic was altered in #466 to strip reference allele content when converting to GenomicLocation to hgvs (for all types except SNP/substitutions which *do* include the replaced nucleotide), and I believe other service layer logic was added when processing hgvs queries so that reference allele specifications are removed before forwarding queries to Ensembl. So this PR is handling the verification of any user specified reference allele string (regardless of backend) through a check of the allele_string member of the returned VariantAnnotation object from whichever backend is configured to make predictions.

Responding to the question of what happens when there is a common prefix:
Users can submit 2 formats : GenomicLocation and HGVS. GenomicLocation always includes both reference allele and variant allele specifications (sometimes '-' for empty). We are accepting HGVS query strings which might or might not contain the reference allele. So in theory we should always be able to pre-process GenomicLocation queries which contain a common prefix (like "AT" -> "AC") and rewrite them as normalized queries which adjust the start position accordingly and issue a query like "T" -> "C" to the VEP backend (whichever one is in use). We will report the predicted variant allele for the normalized query string, but we also include the original_variant_query field to show the raw input. HGVS queries cannot always be pre-processed, because the user may not have provided reference allele string specifiers (for del or delins types). We could (I think we do, but I will check) preprocess HGVS requests to remove common prefixes when the reference allele is provided. When it is not provided, we submit the full query as given to the backend VEP. If a delins "replaces" a prefix substring with an identical replacement variant substring prefix (for example : user query string is 5:g.138163255_138163256delinsTT but the reference assembly sequence in the VEP database for these two positions has sequence "TC" --- having a common prefix of "T")  and then we will receive a response from the backend. This is the case that motivated this work in the first place - the problem is that the local VEP backend will respond with something like "TC->TT" but the public ensembl backend will respond with something like "C->T". This is why this issue was added : #515 --- because in order for these cases to receive a normalized response we would need to post-process or resubmit responses which contain a shared prefix.

In summary : this PR does not modify the preexisting logic which strips common prefix strings from alleles (GenomicLocations have common prefixes removed in NotationConverter.normalizeGenomicLocation(), I am not certain where the HGVS prefix removal code is), so any prefix removal logic should still be applied as before. This PR adds a layer of "reviewing the results" above the AnnotationService classes which were being called before to simply check whether the user's specified reference allele string matches what comes back in the response's "allele_string" data member. (possibly submitting a follow up request if necessary). Both the Ensembl / HGVS annotation server and the genome-nexus-vep / GenomicLocationAnnotationService are covered by the check.